File: 卡码/代码随想录/C++/动态规划/dp/02-背包问题/src/01-01背包/LeetCode_0494.cpp
//494. 目标和
#include <iostream>
#include <vector>
#include <numeric>
#include <algorithm>
using namespace std;

class Solution {
public:
	int findTargetSumWays(vector<int>& nums, int target) {
		int sum=accumulate(nums.begin(),nums.end(),0);
		if(abs(target)>sum||(sum+target)%2==1) return 0;
		int n=(sum+target)/2;
		vector<int> dp(n+1,0);
		dp[0]=1;
		for(int i=0;i<nums.size();i++){
			for(int j=n;j>=nums[i];j--){
				dp[j]+=dp[j-nums[i]];
			}
		}
		return dp[n];
	}
};

int main(){
  system("pause");
  return 0;
}

File: 中等/Java/src/com/kurumi/leetCode/动态规划/LeetCode_0494.java
package com.kurumi.leetCode.动态规划;

/**
 * @Author kurumi
 * @Date 2026-07-06 上午10:54:51
 * @Description 494. 目标和
 */
public class LeetCode_0494 {
    static class Solution {
        public int findTargetSumWays(int[] nums, int target) {
            int sum = 0, n = nums.length;
            for (int i = 0; i < n; i++) {
                sum += nums[i];
            }
            target += sum;
            if (target < 0 || target % 2 != 0) return 0;
            target /= 2;
            int[] dp = new int[target + 1];
            dp[0] = 1;
            for (int num : nums) {
                for (int j = target; j >= num; j--) {
                    dp[j] += dp[j - num];
                }
            }
            return dp[target];
        }
    }

    static class Solution1 {
        public int findTargetSumWays(int[] nums, int target) {
            int sum = 0, n = nums.length;
            for (int i = 0; i < n; i++) {
                sum += nums[i];
            }
            target += sum;
            if (target < 0 || target % 2 != 0) return 0;
            target /= 2;
            int[][] dp = new int[n + 1][target + 1];
            dp[0][0] = 1;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j <= target; j++) {
                    if (j >= nums[i]) dp[i + 1][j] = dp[i][j] + dp[i][j - nums[i]];
                    else dp[i + 1][j] = dp[i][j];
                }
            }
            return dp[n][target];
        }
    }

    public static void main(String[] args) {
        Solution1 solution = new Solution1();
        System.out.println(solution.findTargetSumWays(new int[] {1,1,1,1,1}, 3));
    }
}
